

## Explanation of each set of processed data
 
Unfortunately, the .tsv files at https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE138262 are not the count tables that were used in the paper Wehrens et al. (2022), Cell Reports (https://doi.org/10.1016/j.celrep.2022.110809). 

The following files contain the correct data, and additionally provide some further processed data files (such as seurat objects).
 
- The data in *counts.table.tsv (see list below) contain data processed by the pipeline described in Wehrens et al. and include:
    1. the read counts of RNA-seq data of hearts of humans with HCM, that were generated by Wehrens et al.
        Samples: See records GSM4103885-GSM4762108 below.
    2. re-generated count tables, based on FASTQ data published earlier by Wang et al., related to records GSE121893 and GSE109816.
        Reads were selected from the data based on their tissue location and cluster assignments, as described in the methods section of Wehrens et al.
        The samples include: 
        N1_LV (GSM2970361), N2_LV (GSM2970358), N3_LV (GSM2970362), N4_LV (GSM2970366), N5_LV (GSM2970360), N13 (GSM3449619) and N14 (GSM3449620)
- Index data from FACS-sorting, as also available in GSM4103897, GSM4103898, GSM4762107, GSM4762108, as well as an Rdata file with that data re-organized. This data was used for cell size analysis as described in Wehrens et al.
- The data was further analyzed using Seurat. Seurat objects are available in the .Rds files, for both
    (1) data specifically generated by Wehrens et al., and
        File: H5_RHL_SeuratObject_nM_sel_ROOIJonly.sp.bt_RID2l_clExtended.Rds
    (2) data generated by Wehrens et al. in comparison to data generated by other researchers, as described in the paper (Fig 1), extracted from GSE121893, GSE109816 and ERP123138
        File: H5_RHL_SeuratObject_nM_sel_ALL.SP_btypSel_RID2l_clExtended.Rds

(The current records for patients 4 and 5 also contain index data, e.g. GSM4762108_indexdata_patient5-plate2_AL2.csv.gz, which is still valid.)
 
 ## List of each new processed data file with the corresponding Sample (GSMxxx)
 
- Data from samples processed by Wehrens et al. (processed as described in the methods section:

| GEO | Patient | Plate | Filename |
|-----|---------|--------|----------|
|  GSM4103885  |  Patient 1 | Plate 1  |  rJE1_AHFL7NBGX5_S3_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103886  |  Patient 1 | Plate 2  |  JE2_AHY3WGBGX3_S1_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103887  |  Patient 1 | Plate 3  |  JE3_AHY3WGBGX3_S2_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103888  |  Patient 1 | Plate 4  |  JE4_AHFL7NBGX5_S4_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103889  |  Patient 2 | Plate 1   |  JE5_AHFL77BGX5_S6_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103890  |  Patient 2 | Plate 2   |  JE6_AHFL77BGX5_S7_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103891  |  Patient 2 | Plate 3   |  JE7_AHFL7NBGX5_S16_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103892  |  Patient 2 | Plate 4   |  JE8_AHFL7NBGX5_S17_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103893  |  Patient 3 | Plate 1  |  HUB-MW-005_AH32W2BGX9_S5_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103894  |  Patient 3 | Plate 2  |  HUB-MW-006_AH32W2BGX9_S6_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103895  |  Patient 3 | Plate 3  |  HUB-MW-007_HC3GFBGX9_S6_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103896  |  Patient 3 | Plate 4  |  HUB-MW-008_HC3GFBGX9_S7_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103897  |  Patient 4 | Plate 1  |  HUB-JE-010_HGVN3BGX9_S1_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4103898  |  Patient 4 | Plate 2  |  HUB-JE-011_HGVN3BGX9_S2_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4762107  |  Patient 5 | Plate 1  |  HUB-AL-s001_HG25TBGXF_S5_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|  GSM4762108  |  Patient 5 | Plate 2  |  HUB-AL-s002_HG25TBGXF_S6_cat_pT.nonRibo_E99_Aligned.out.counts.table.tsv  |
|-----|-------|----------|----------|


File: H5_RHL_SeuratObject_nM_sel_ROOIJonly.sp.bt_RID2l_clExtended.Rds
A Seurat object with only the processed data from Wehrens et al samples, samples:
```
GSM4103885, GSM4103886, GSM4103887, GSM4103888, GSM4103889, GSM4103890, GSM4103891, GSM4103892, 
GSM4103893, GSM4103894, GSM4103895, GSM4103896, GSM4103897, GSM4103898, GSM4762107, GSM4762108.
```

- Additional  files: index data from FACS sorting.
```
GSM4103897    GSM4103897_indexdata_patient4-plate1__2018-244_exp_cardiologie_plate2.csv
GSM4103898    GSM4103898_indexdata_patient4-plate2__2018-244_exp_cardiologie_plate3merged.csv
GSM4762107    GSM4762107_indexdata_patient5-plate1_AL1.csv
GSM4762108    GSM4762108_indexdata_patient5-plate2_AL2.csv
```

Data that is already available at GSM4103897, related to the mentioned records.

File:         FSCA__indexData_MW.Rdata
Combined post-processed index sorted data above in Rdata format, containing the table indexData_MW with columns 'FSC_A','FSC_W','FSC_H','SSC_A','SSC_W','SSC_H' and the cellular ID.
  Contains data from: GSM4103897, GSM4103898, GSM4762107, GSM4762108

- Data from Wang et al (re-calculated count tables determined by Wehrens et al. paper from selected data as described in the Wehrens et al. methods section):

|-------------|--------------------------------------------------------------------------|
| GEO id      | Filename                                                                 |  
|-------------|--------------------------------------------------------------------------|
| GSM2970361  |  p.N1.plate.97474.part.1_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970361  |  p.N1.plate.97474.part.2_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970358  |  p.N2.plate.97452.part.1_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970358  |  p.N2.plate.97452.part.2_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970358  |  p.N2.plate.97493.part.1_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970358  |  p.N2.plate.97493.part.2_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970362  |  p.N3.plate.97438.part.1_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970362  |  p.N3.plate.97438.part.2_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970366  |  p.N4.plate.97461.part.1_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970366  |  p.N4.plate.97461.part.2_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970360  |  p.N5.plate.97458.part.1_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM2970360  |  p.N5.plate.97458.part.2_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM3449619  |  p.N13.plate.100355.part.1_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM3449619  |  p.N13.plate.100355.part.2_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM3449620  |  p.N14.plate.104720.part.1_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
| GSM3449620  |  p.N14.plate.104720.part.2_cat_nc.nonRibo_E99_Aligned.out.counts.table.tsv |
|-------------|--------------------------------------------------------------------------|

 
File: H5_RHL_SeuratObject_nM_sel_ALL.SP_btypSel_RID2l_clExtended.Rds
This file contains data processed using Seurat as described in Wehrens et al., Wang et al., and Litvinukova et al. Data processing is described in the methods section. 
This file contains data from:
```
GSM4103885, GSM4103886, GSM4103887, GSM4103888, GSM4103889, GSM4103890, GSM4103891, GSM4103892, GSM4103893, GSM4103894, GSM4103895, GSM4103896, GSM4103897, GSM4103898, GSM4762107, GSM4762108.
GSM2970361, GSM2970361, GSM2970358, GSM2970358, GSM2970358, GSM2970358, GSM2970362, GSM2970362, GSM2970366, GSM2970366, GSM2970360, GSM2970360, GSM3449619, GSM3449619, GSM3449620, GSM3449620  
ERP123138
```
  
 ## Description of the contents of the new files
 
 See also above.
 
 *counts.table.tsv    contain read counts (UMI counts) of RNA-seq data in tsv format.
 *.Rds                contain R objects generated by the Seurat library.
 *.Rdata              contain objects generated by R with tabular data.
 
 ## Data processing steps used to generate the new processed data, including reference genome/assembly
 
 Reference genome: GRCh38.93
 
 Methods (as published in Wehrens et al. 2022):
 
 (1)
 Mapping of HCM sequencing data  
 (Applies to samples GSM4103885, GSM4103886, GSM4103887, GSM4103888, GSM4103889, GSM4103890, GSM4103891, GSM4103892, GSM4103893, GSM4103894, GSM4103895, GSM4103896, GSM4103897, GSM4103898, GSM4762107, GSM4762108.)
 After sequencing, each of the HCM samples were mapped. Read 1 (R1) contains respectively the UMI (positions 1–6) and barcode (positions 7–14) followed by the poly-T sequence (theoretically 24 nt) and sample mRNA sequence, read 2 (R2) contains the mRNA transcript sequence only. UMI and cell barcode (BC) information was attached to R2 metadata, and only R2 was used for mapping. In addition, we discarded reads that did not show R10 thymine nucleotides after the BC in R1. Reads were then trimmed for sequencing primers and repeating single nucleotides using the Cutadapt (Martin, 2011) wrapper Trim Galore (default settings). We then filtered out ribosomal RNA reads by discarding reads that mapped to the human rRNA genes using Burrows-Wheeler Alignment Tool (bwa merge of bwa aln and bwa mem -h 15) (Li and Durbin, 2009). These pre-processed reads were then mapped using Spliced Transcripts Alignment to a Reference (STAR –outFilterMultimapNmax 20) (Dobin et al., 2013), assigned to genes using featureCounts (-R BAM) (Liao et al., 2014), where multi-mappers were ignored (featureCounts default behavior), and processed into UMI count tables using UMI-tools (umi_tools count; default settings) (Smith et al., 2017). As reference genome, we used the human genome (GRCh38.93) and gene annotation (gtf file) acquired from ensembl.org. The gtf file was filtered to only contain genes with a gene_biotype annotation of protein_coding, lincRNA or antisense. After mapping, HCGN symbols were linked to Ensembl IDs using the R biomaRt package.
 
 (2)
 Obtaining Ctrl1 and Ctrl2 count tables  
 (Applies to samples GSM2970361, GSM2970361, GSM2970358, GSM2970358, GSM2970358, GSM2970358, GSM2970362, GSM2970362, GSM2970366, GSM2970366, GSM2970360, GSM2970360, GSM3449619, GSM3449619, GSM3449620, GSM3449620.)
 (Count tables from Litvinukova/ERP123138 were used directly, as described below.)
 We included cells from healthy donor hearts from two different studies, Wang et al. (Wang et al., 2020) (referred to as Ctrl1) and Litvinukova ́ et al. (Litvinukova ́ et al., 2020) (Ctrl2). From Wang et al., we used samples N1, N2, N3, N4, N5, N13 and N14, as they came from healthy donors and included LV cells. We obtained raw FASTQ files from Gene Expression Omnibus (GEO, accession numbers GSE121893 and GSE109816) using the NCBI SRA Toolkit (fasterq-dump command) and SRR identifiers obtained from the provided metadata files. We collected FASTQ data for cells annotated as LV cells (clusters LV1-9) into files per donor and plate (we note that these selected cells were filtered for quality already by Wang et al.). These data were collected with the iCell8 platform from Takara Bio, R2 again contained the biological information that was mapped, and R1 contained only the BC (11 nt) followed by the UMI (ranging from 10-14 nt), of which we used only the first 10 nucleotides to deal with the variable length. (A reference BC file was obtained from the Cogent NGS Analysis Pipeline, v. 1.0, Takara Bio.) We mapped the obtained FASTQ files using the same pipeline as the HCM samples. For Litvinukova ́ et al. (who used 103 Genomics) we downloaded count tables for CMs for all available samples (D1-7, D11, H2-7) from https://www.heartcellatlas.org/ (h5ad files) and selected only cells taken from the septum. These tables were generated by the Cell Ranger pipeline (v. 3.0.2, 103 Genomics), which, identical to our pipeline, uses the STAR aligner in combination with human genome GRCh38.93 and gene annotation (ensembl.org), and only considers uniquely mapped reads (103 Genomics documentation). The downloaded Litvinukova ́ et al. raw count tables were converted to h5seurat format and used directly for analysis in Seurat.


 
 
 